File: one-more-time/src/cpp/one-more-time.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
    int target_degree;
    cout << "Input target degree: ";
    cin >> target_degree;

    vector<double> x_points;
    vector<double> y_points;

    cout << "Input xi yi" << endl;
    for (int i = 0; i <= target_degree; i++) {
        double xi, yi;
        cin >> xi >> yi;
        x_points.push_back(xi);
        y_points.push_back(yi);
    }

    vector<double> lagrange_sums;
    for (int i = 0; i < x_points.size(); i++) {
        double constant = y_points[i];

        vector<double> lagrange_terms;
        for (int j = 0; j < x_points.size(); j++) {
            if (i != j) {
                constant /= x_points[i] - x_points[j];
                if (lagrange_terms.empty()) {
                    lagrange_terms.push_back(1);
                    lagrange_terms.push_back(-x_points[j]);
                } else {
                    vector<double> temp(lagrange_terms);
                    lagrange_terms.push_back(0);

                    for (int k = 0; k < temp.size(); k++) {
                        temp[k] *= -x_points[j];
                        lagrange_terms[k + 1] += temp[k];
                    }
                }
            }
        }
        
        transform(lagrange_terms.begin(), lagrange_terms.end(), lagrange_terms.begin(), 
                  [constant](double val) { return val * constant; });

        if (lagrange_sums.empty()) {
            lagrange_sums = vector<double>(lagrange_terms);
        } else {
            transform(lagrange_sums.begin(), lagrange_sums.end(), lagrange_terms.begin(), lagrange_sums.begin(), plus<double>());
        }
    }
    
    int degree = target_degree;
    for (double term : lagrange_sums) {
        if (term != 0) {
            cout << "(" << term << ")";
            if (degree != 0) {
                cout << "X^" << degree << " + ";
            }
        }
        degree--;
    }
    cout << endl;
    
    return 0;
}

File: one-more-time/src/python-3.x/one-more-time.py
def main():
    target_degree = int(input("Target degree: "))
    x_points = []
    y_points = []

    # input
    for i in range(target_degree + 1):
        xi = float(input("x%d = " % i))
        yi = float(input("y%d = " % i))

        x_points.append(xi)
        y_points.append(yi)

    lagrange_sums = []
    for i in range(len(x_points)):
        constant = y_points[i]

        lagrange_terms = []
        for j in range(len(x_points)):
            if i == j:
                continue
            else:
                constant /= x_points[i] - x_points[j]
                if len(lagrange_terms) == 0:
                    lagrange_terms = [1, - x_points[j]]
                else:
                    temp = list(lagrange_terms)
                    lagrange_terms.append(0)

                    for k in range(len(temp)):
                        temp[k] *= - x_points[j]
                        lagrange_terms[k + 1] += temp[k]

        lagrange_terms = map(lambda x: x * constant, lagrange_terms)
        if len(lagrange_sums) == 0:
            lagrange_sums = list(lagrange_terms)
        else:
            lagrange_sums = [sum(i) for i in zip(lagrange_sums, lagrange_terms)]
    

    for i, a in enumerate(lagrange_sums):
        if a != 0:
            print(f"({a})", end="")

            if target_degree - i != 0:
                print(f"X^{target_degree - i} + ", end="")
    print()




        


if __name__ == "__main__":
    main()